Fix inaccuracies in vis_heap_chunks()

CptGibbon opened the pull request:
The `vis_heap_chunks` command incorrectly determines the boundaries between chunks near to a top chunk. For example, in the image below there are 3 identical, allocated chunks of minimum size which `vis_heap_chunks` wrongly shows as having different shapes. The 2nd invocation has not printed the trailing quadword of the 1st chunk, despite it being correctly marked as belonging to the chunk in the 1st output.

<img width="1044" alt="image" src="https://user-images.githubusercontent.com/16000770/74738894-b1606500-524f-11ea-89c6-248df8d73f11.png">

This PR corrects the boundary calculation of chunks adjacent to a top chunk, it also uses a list of boundary addresses to do so, rather than a dictionary of every quadword in each printed chunk. The image below shows `vis_heap_chunks` output after the patch has been applied, each chunk looks identical and the 2nd invocation prints the last quadword of the 1st chunk.

<img width="1044" alt="image" src="https://user-images.githubusercontent.com/16000770/74739303-7579cf80-5250-11ea-8a6e-e8f27a208573.png">

The colours change between screenshots because the original implementation subtracted 1 from a `color_func_idx` variable that was already 0 before building the output buffer. Fixes #706

CptGibbon committed:
Fix inaccuracies in vis_heap_chunks()

Correct boundary calculation of chunks adjacent to a top chunk and
use a list of boundary addresses to do so, rather than a dictionary of
every quadword in each printed chunk. Ensure that at least 1 chunk is
printed even when a non-sensical value of 0 is passed as the count argument.

CptGibbon committed:
Avoid out-of-bounds read in vis_heap_chunks()

While processing a corrupt heap or using the --naive option, the
vis_heap_chunks command could read out-of-bounds. Add a check to
avoid this scenario.

disconnect3d commented:
@GrosQuildu or @stnevans any chance one of u could review this?

GrosQuildu commented on pwndbg/commands/heap.py:
```diff
@@ -495,51 +495,51 @@ def find_fake_fast(addr, size):
+
+        size_field = pwndbg.memory.u(cursor + ptr_size)
+        real_size = size_field & ~main_heap.malloc_align_mask
+        prev_inuse = size_field & pwndbg.constants.ptmalloc.PREV_INUSE
```
```suggestion
        prev_inuse = main_heap.chunk_flags(size_field)[0]
```

GrosQuildu commented on pwndbg/commands/heap.py:
```diff
@@ -495,51 +495,51 @@ def find_fake_fast(addr, size):
 
-    cells_map = {}
-    chunk_id = 0
     ptr_size = pwndbg.arch.ptrsize
```
Just for quality, not really important.
```suggestion
    ptr_size = main_heap.size_sz
```

GrosQuildu commented on pwndbg/commands/heap.py:
```diff
@@ -495,51 +495,51 @@ def find_fake_fast(addr, size):
+        else:
+            chunk_delims.append(cursor)
+
+        if (cursor == top_chunk and not naive) or (cursor == heap_region.end - 0x10):
```
0x10 - magic constant, maybe better will be:
```suggestion
        if (cursor == top_chunk and not naive) or (cursor == heap_region.end - ptr_size*2):
```

GrosQuildu commented on pwndbg/commands/heap.py:
```diff
@@ -563,30 +562,33 @@ def vis_heap_chunks(address=None, count=None, naive=None):
+
+    for c, stop in enumerate(chunk_delims):
+        color_func_idx = c % len(color_funcs)
+        color_func = color_funcs[color_func_idx]
```
```suggestion
        color_func = color_funcs[c % len(color_funcs)]
```

GrosQuildu commented on pwndbg/commands/heap.py:
```diff
@@ -563,30 +562,33 @@ def vis_heap_chunks(address=None, count=None, naive=None):
-        cell = unpack(pwndbg.memory.read(addr, ptr_size))
-        cell_hex = '\t0x{:0{n}x}'.format(cell, n=ptr_size*2)
+            cell = pwndbg.memory.u(cursor)
+            cell_hex = '\t0x{:0{n}x}'.format(cell, n=ptr_size*2)
```
Does pwndbg have some uniform way to print stuff like that? That is: address, memory, content, and labels, kinda like hexdump. @disconnect3d 

Probably not?

disconnect3d replied:
I think not

GrosQuildu commented:
Can we make it works with non-heap regions? Running `vis_heap_chunks` on arbitrary address may be useful when you are constructing fake chunks on i.e. stack. Currently its not possible because [`main_heap.get_heap_boundaries(address)`](https://github.com/pwndbg/pwndbg/blob/7f963fed0e8f54de93101a33ce30e66d497fb058/pwndbg/commands/heap.py#L508) crashes.

disconnect3d commented:
@CptGibbon ping :)

CptGibbon commented:
> @CptGibbon ping :)

I'm on it 👍

CptGibbon committed:
Use chunk_flags instead of constants

To find out whether a chunk's prev_inuse bit is set, use chunk_flags() rather than anding with constants.ptmalloc.PREV_INUSE.

Co-Authored-By: Paweł Płatek <e2.8a.95@gmail.com>

CptGibbon committed:
Use main_heap.size_sz

Use main_heap.size_sz in place of pwndbg.arch.ptrsize.

Co-Authored-By: Paweł Płatek <e2.8a.95@gmail.com>

CptGibbon committed:
Replace a magic number with ptr_size

Co-Authored-By: Paweł Płatek <e2.8a.95@gmail.com>

CptGibbon committed:
Calculate color_func variable in 1 line

Co-Authored-By: Paweł Płatek <e2.8a.95@gmail.com>

CptGibbon committed:
Merge branch 'dev' into fix-vis_heap_chunks

CptGibbon committed:
Fix crash in get_heap_boundaries()

Don't assume that because an address was supplied to
get_heap_boundaries() that it resides on a legitimate heap. Sometimes a
user may want to print fake chunks elsewhere in memory.

CptGibbon commented:
> Can we make it works with non-heap regions? Running `vis_heap_chunks` on arbitrary address may be useful when you are constructing fake chunks on i.e. stack. Currently its not possible because [`main_heap.get_heap_boundaries(address)`](https://github.com/pwndbg/pwndbg/blob/7f963fed0e8f54de93101a33ce30e66d497fb058/pwndbg/commands/heap.py#L508) crashes.

Sure, I've addressed this in 62f1180.

GrosQuildu commented:
Seems ok, thanks.

@disconnect3d can merge that one.

GrosQuildu reviewed (approved):
OK
